### PR TITLE
Add Batch Changes methods to Gerrit Client

### PR DESCRIPTION
**Pull request by varsanojidan**
Closes https://github.com/sourcegraph/sourcegraph/issues/50866

Gerrit is a bit weird so the creation and updating portions of Changes are done through commits/pushes and therefore don't need API methods.

This probably isn't a comprehensive list for everything needed but I will just add whatever methods are missing to followup tickets,

Don't be scared, this is a lot of generated code.

## Test plan
Added unit tests to cover the new methods.

### CODE REVIEW

**Comment by varsanojidan** [comment]
Looping in the IAM team because I made slight changes to the Gerrit auth provider/test, don't worry about the rest of the changes.

File: enterprise/internal/authz/gerrit/gerrit.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ const (
 
 type Provider struct {
 	urn      string
-	client   client
+	client   gerrit.Client
```
Comment (kopancek) [inline enterprise/internal/authz/gerrit/gerrit.go:21]: Can you explain to the uninitiated why this change was needed?

File: enterprise/internal/authz/gerrit/gerrit.go [hunk 2]
```diff
@@ -62,10 +62,13 @@ func (p Provider) FetchUserPerms(ctx context.Context, account *extsvc.Account, o
 		return nil, err
 	}
 
-	client := p.client.WithAuthenticator(&auth.BasicAuth{
+	client, err := p.client.WithAuthenticator(&auth.BasicAuth{
```
Comment (kopancek) [inline enterprise/internal/authz/gerrit/gerrit.go:65]: nice, error handling FTW!

File: internal/extsvc/gerrit/changes.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+
+	"github.com/sourcegraph/sourcegraph/lib/errors"
+)
+
+func (c *client) GetChange(ctx context.Context, changeID string) (*Change, error) {
+	reqURL := url.URL{Path: fmt.Sprintf("a/changes/%s", changeID)}
```
Comment (kopancek) [inline internal/extsvc/gerrit/changes.go:15]: We should use `PathEscape` on the `changeID`, since it's a string. Just in case. Also consider using [`url.JoinPath`](https://pkg.go.dev/net/url#JoinPath) instead of `fmt.Sprintf` here.

File: internal/extsvc/gerrit/changes.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	return &change, nil
+}
+
+// AbandonChange abandons a Gerrit change.
+func (c *client) AbandonChange(ctx context.Context, changeID string) (*Change, error) {
+	reqURL := url.URL{Path: fmt.Sprintf("a/changes/%s/abandon", changeID)}
```
Comment (kopancek) [inline internal/extsvc/gerrit/changes.go:36]: Same as above

File: internal/extsvc/gerrit/changes.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	return &change, nil
+}
+
+// SubmitChange submits a Gerrit change.
+func (c *client) SubmitChange(ctx context.Context, changeID string) (*Change, error) {
+	reqURL := url.URL{Path: fmt.Sprintf("a/changes/%s/submit", changeID)}
```
Comment (kopancek) [inline internal/extsvc/gerrit/changes.go:57]: At this point I'm thinking if we need a standard sg library for merging url paths? 😅

File: internal/extsvc/gerrit/changes.go [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+	if resp.StatusCode != http.StatusOK {
+		return nil, errors.Errorf("unexpected status code: %d", resp.StatusCode)
+	}
+
+	return &change, nil
+}
+
+// WriteReviewComment writes a review comment on a Gerrit change.
+func (c *client) WriteReviewComment(ctx context.Context, changeID string, comment ChangeReviewComment) error {
+	reqURL := url.URL{Path: fmt.Sprintf("a/changes/%s/revisions/current/review", changeID)}
```
Comment (kopancek) [inline internal/extsvc/gerrit/changes.go:78]: Same old, as above.

File: internal/extsvc/gerrit/changes.go [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+		return nil, err
+	}
+
+	var change Change
+	resp, err := c.do(ctx, req, &change)
+	if err != nil {
+		return nil, err
+	}
+
+	if resp.StatusCode != http.StatusOK {
```
Comment (kopancek) [inline internal/extsvc/gerrit/changes.go:27]: Does this always end up being a 200 on success? Can it return weird 2xx codes or a redirect? Should we only return error if it's > 400? Not sure how Gerrit API works.

File: internal/extsvc/gerrit/client.go [hunk 8]
```diff
@@ -45,24 +56,31 @@ func NewClient(urn string, url *url.URL, creds *AccountCredentials, httpClient h
...
 
-func (c *Client) WithAuthenticator(a auth.Authenticator) *Client {
-	return &Client{
+func (c *client) WithAuthenticator(a auth.Authenticator) (Client, error) {
+	switch a.(type) {
+	case *auth.BasicAuth, *auth.BasicAuthWithSSH:
+		break
+	default:
+		return nil, errors.Errorf("authenticator type unsupported for Azure DevOps clients: %s", a)
+	}
```
Comment (kopancek) [inline internal/extsvc/gerrit/client.go:73]: Nit: this might be just me, but since this is basically returning an error for unsupported cases, I would not use a switch, but a plain if condition instead. Seems much shorter to me and more readable.

```suggestion
	if a.(type) != *auth.BasicAuth && a.(type) != *auth.BasicAuthWithSSH {
		return nil, errors.Errorf("authenticator type unsupported for Azure DevOps clients: %s", a)
	}
```

File: internal/extsvc/gerrit/projects.go [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+
+func (c *client) listCodeProjects(ctx context.Context, cursor *Pagination) (ListProjectsResponse, bool, error) {
+	// Unfortunately Gerrit APIs are quite limited and don't support pagination well.
+	// e.g. when we request a list of 100 CODE projects, 100 projects are fetched and
+	// only then filtered for CODE projects, possibly returning less than 100 projects.
+	// This means we cannot rely on the number of projects returned to determine if
+	// there are more projects to fetch.
+	// Currently, if you want to only get CODE projects and want to know if there is another page
+	// to query for, the only way to do that is to query both CODE and ALL projects and compare
+	// the number of projects returned.
```
Comment (kopancek) [inline internal/extsvc/gerrit/projects.go:18]: This really really sucks.

Another way would be to filter on our side. Just get ALL projects and filter for CODE projects. That way we can save ourselves 1 last request if we get no projects as a result and we know that there is no next page.

Seems to me like that logic would be easier to understand.

File: internal/extsvc/gerrit/changes.go (hunk 7 above)
Comment (varsanojidan) [reply internal/extsvc/gerrit/changes.go:27]: Yeah its only 200 on success, but its probably a bit safer to go with >=400, just in case something changes bw versions.

File: internal/extsvc/gerrit/client.go (hunk 8 above)
Comment (varsanojidan) [reply internal/extsvc/gerrit/client.go:73]: Unfortunately, a.(type) only works within switch statements, so if you wanted to avoid the switch block you would need to check each type in a separate if statement, and I think the switch is a bit cleaner.

File: internal/extsvc/gerrit/projects.go (hunk 9 above)
Comment (varsanojidan) [reply internal/extsvc/gerrit/projects.go:18]: That would be amazing, but unfortunately, when you query for projects, Gerrit does not tell you if they are code projects or not 🙃, so you have to cross-reference it to API request that queries only for code projects. 

This is a sample output from the query to get projects:


```
)
]
}'
{
"All-Projects": {
"id": "All-Projects",
"state": "ACTIVE"
},
"All-Users": {
"id": "All-Users",
"state": "ACTIVE"
},
"beatrix": {
"id": "beatrix",
"state": "ACTIVE"
},
```

File: internal/extsvc/gerrit/client.go (hunk 8 above)
Comment (kopancek) [reply internal/extsvc/gerrit/client.go:73]: TIL. thank you for the explanation 👍

**Review by kopancek — APPROVED** [review]

File: internal/extsvc/gerrit/changes_test.go [hunk 10]
```diff
@@ -0,0 +1,63 @@
+package gerrit
+
+import (
+	"context"
+	"testing"
+
+	"github.com/sourcegraph/sourcegraph/internal/testutil"
+)
+
+func TestClient_GetChange(t *testing.T) {
```
Comment (courier-new) [inline internal/extsvc/gerrit/changes_test.go:10]: Not urgent, but it could be helpful to leave some inline documentation here about which repo and change each of these VCR tests is looking for, plus URLs. It's also extremely helpful to list out any pre-conditions required in order to re-record the test (e.g. "the change used for this test will be closed by it. before re-running, it must be re-opened" or "the change used for this test was already merged. to re-record this test, push a new branch with at least one commit, then find and update the change ID"), so that the next time someone else comes through and needs to re-run them in a pinch, they're not having to start from scratch. 🙂

File: internal/extsvc/gerrit/projects_test.go [hunk 11]
```diff
@@ -0,0 +1,26 @@
+package gerrit
+
+import (
+	"context"
+	"testing"
+
+	"github.com/sourcegraph/sourcegraph/internal/testutil"
+)
+
+func TestClient_ListProjects(t *testing.T) {
```
Comment (courier-new) [inline internal/extsvc/gerrit/projects_test.go:10]: Same for this test.

File: internal/extsvc/gerrit/changes_test.go (hunk 10 above)
Comment (varsanojidan) [reply internal/extsvc/gerrit/changes_test.go:10]: Yeah for sure, I’ll add it to my list to add in the future PRs 🙂
